# **Netflix Movies and TV Shows - Data Cleaning**

## **Часть 1. Введение**

Данный датасет содержит информацию о фильмах и сериалах, доступных на Netflix.
В сыром виде данные содержат пропуски в колонках, смешанные форматы и даты, которые хранятся в виде строк.

Целью проекта является провести осмысленную очистку данных: разобраться в природе пропусков по каждой колонке и принять обоснованное решение по ней, привести типы данных к корректным, и подготовить датасет к дальнейшему анализу.

## **Часть 2. Загрузка и первый обзор данных**

In [ ]:
import pandas as pd
import os

In [2]:
os.chdir('..') 
df = pd.read_csv('data/raw/netflix_titles.csv')

In [3]:
print(f"Датафрейм содержит {df.shape[0]} строк и {df.shape[1]} колонок")

Датафрейм содержит 8807 строк и 12 колонок


In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   show_id       8807 non-null   str  
 1   type          8807 non-null   str  
 2   title         8807 non-null   str  
 3   director      6173 non-null   str  
 4   cast          7982 non-null   str  
 5   country       7976 non-null   str  
 6   date_added    8797 non-null   str  
 7   release_year  8807 non-null   int64
 8   rating        8803 non-null   str  
 9   duration      8804 non-null   str  
 10  listed_in     8807 non-null   str  
 11  description   8807 non-null   str  
dtypes: int64(1), str(11)
memory usage: 825.8 KB


In [5]:
missing = df.isna().sum()  # количество пропущенных значений по каждому столбцу
missing.sort_values(ascending=False)

director        2634
country          831
cast             825
date_added        10
rating             4
duration           3
show_id            0
type               0
title              0
release_year       0
listed_in          0
description        0
dtype: int64

In [6]:
round(missing / len(df) * 100, 2).sort_values(ascending=False)  # процент пропусков по каждому столбцу

director        29.91
country          9.44
cast             9.37
date_added       0.11
rating           0.05
duration         0.03
show_id          0.00
type             0.00
title            0.00
release_year     0.00
listed_in        0.00
description      0.00
dtype: float64

In [7]:
if df.duplicated().sum() == 0:
    print("Полностью совпадающих строк нет")
else:
    print(f"Найдено {len(df[df.duplicated()])} дубликатов")
    df[df.duplicated()]

Полностью совпадающих строк нет


In [8]:
df.head(10)

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,NaN,United States,"September 25, 2021",2020,PG-13,90 min,Documentaries,"As her father nears the end of his life, filmm..."
1,s2,TV Show,Blood & Water,NaN,"Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...",South Africa,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, TV Dramas, TV Mysteries","After crossing paths at a party, a Cape Town t..."
2,s3,TV Show,Ganglands,Julien Leclercq,"Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...",NaN,"September 24, 2021",2021,TV-MA,1 Season,"Crime TV Shows, International TV Shows, TV Act...",To protect his family from a powerful drug lor...
3,s4,TV Show,Jailbirds New Orleans,NaN,NaN,NaN,"September 24, 2021",2021,TV-MA,1 Season,"Docuseries, Reality TV","Feuds, flirtations and toilet talk go down amo..."
4,s5,TV Show,Kota Factory,NaN,"Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...",India,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, Romantic TV Shows, TV ...",In a city of coaching centers known to train I...
5,s6,TV Show,Midnight Mass,Mike Flanagan,"Kate Siegel, Zach Gilford, Hamish Linklater, H...",NaN,"September 24, 2021",2021,TV-MA,1 Season,"TV Dramas, TV Horror, TV Mysteries",The arrival of a charismatic young priest brin...
6,s7,Movie,My Little Pony: A New Generation,"Robert Cullen, José Luis Ucha","Vanessa Hudgens, Kimiko Glenn, James Marsden, ...",NaN,"September 24, 2021",2021,PG,91 min,Children & Family Movies,Equestria's divided. But a bright-eyed hero be...
7,s8,Movie,Sankofa,Haile Gerima,"Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...","United States, Ghana, Burkina Faso, United Kin...","September 24, 2021",1993,TV-MA,125 min,"Dramas, Independent Movies, International Movies","On a photo shoot in Ghana, an American model s..."
8,s9,TV Show,The Great British Baking Show,Andy Devonshire,"Mel Giedroyc, Sue Perkins, Mary Berry, Paul Ho...",United Kingdom,"September 24, 2021",2021,TV-14,9 Seasons,"British TV Shows, Reality TV",A talented batch of amateur bakers face off in...
9,s10,Movie,The Starling,Theodore Melfi,"Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...",United States,"September 24, 2021",2021,PG-13,104 min,"Comedies, Dramas",A woman adjusting to life after a loss contend...


In [9]:
df.describe()

,release_year
count,8807.000000
mean,2014.180198
std,8.819312
min,1925.000000
25%,2013.000000
50%,2017.000000
75%,2019.000000
max,2021.000000


In [10]:
columns = list(df.columns)
columns

['show_id',
 'type',
 'title',
 'director',
 'cast',
 'country',
 'date_added',
 'release_year',
 'rating',
 'duration',
 'listed_in',
 'description']

In [11]:
# Количество уникальных значений по столбцам
for column in columns:
    print(f"{column}: {df[column].nunique()}")

show_id: 8807
type: 2
title: 8807
director: 4528
cast: 7692
country: 748
date_added: 1767
release_year: 74
rating: 17
duration: 220
listed_in: 514
description: 8775


In [12]:
# Для просмотра уникальных значений нужно раскомментировать код ниже

# for column in columns:
#     pprint(f"{column}: {df[column].unique().tolist()}")
#     print('\n')

### **Выводы после первичного осмотра:**

В таблице всего 8807 строк и 12 столбцов. Полностью совпадающих строк нет.

Пропуски присутствуют в 6 колонках: в колонке `director` 2634 пропусков, что составляет 29.91% (почти треть столбца), в `country` - 831 (9.44%), `cast` - 825 (9.37%), `date_added` - 10 (0.11%), `rating` - 4 (0.05%), `duration` - 3 (0.03%). 

В столбцах без пропусков никаких аномалий нет.



## **Часть 3. Анализ и обработка пропусков**

### **3.1 director**

В столбце `director` аномалий нет, но пропущено около 30% значений, что составляет почти треть. Удалять данный столбец не стоит, потому что мы потеряем более 2.5 тыс. строк, а в этих записях может быть нужная информация для анализа по другим метрикам.

In [13]:
df.groupby('type')['director'].apply(lambda x: x.isna().mean() * 100)

type
Movie       3.066384
TV Show    91.405082
Name: director, dtype: float64

В ячейке выше мы посчитали процент пропусков по каждому типу. Выяснилось, что у фильмов всего 3% пропусков (данные почти полные, поэтому по ним можно проводить аналитику), а у ТВ-шоу - 91%.

И это неудивительно, потому что режиссеров обычно не указывают для ТВ-шоу. Поэтому все пропуски в столбце `director` мы заполним значением "Unknown", данный столбец удалять не нужно.

In [14]:
df['director'] = df['director'].fillna('Unknown')

### **3.2 cast**

In [15]:
df[df['cast'].isna()]['listed_in'].value_counts().head(15)

listed_in
Documentaries                                            183
Documentaries, International Movies                      117
Docuseries                                                47
Crime TV Shows, Docuseries                                36
Reality TV                                                31
Documentaries, Sports Movies                              30
Crime TV Shows, Docuseries, International TV Shows        24
Kids' TV                                                  23
Documentaries, International Movies, Sports Movies        21
Documentaries, Music & Musicals                           20
Docuseries, Science & Nature TV                           18
Docuseries, International TV Shows                        17
Children & Family Movies                                  16
Docuseries, Reality TV                                    12
Documentaries, International Movies, Music & Musicals     12
Name: count, dtype: int64

Мы видим, что пропуски по большей части в документальных фильмах/сериалах, в них актеров обычно не указывают.

Аномалий в `cast` нет, поэтому пустые значения заполним "Unknown".

In [16]:
df['cast'] = df['cast'].fillna("Unknown")

### **3.3 country**

Пустые значения заполним "Unknown", а так же уберем лишние пробелы с запятыми в значениях `, South Korea` и `, France, Algeria`

In [17]:
df['country'] = df['country'].str.strip(', ')

In [18]:
df['country'] = df['country'].fillna("Unknown")

### **3.4 rating**

In [19]:
list(df['rating'].unique())

['PG-13',
 'TV-MA',
 'PG',
 'TV-14',
 'TV-PG',
 'TV-Y',
 'TV-Y7',
 'R',
 'TV-G',
 'G',
 'NC-17',
 '74 min',
 '84 min',
 '66 min',
 'NR',
 nan,
 'TV-Y7-FV',
 'UR']

Если мы выведем все уникальные значения рентинга, то увидим аномалию `['74 min', '84 min', '66 min']`.

Вставим значения "NR" вместо пустых ячеек, а также заменим на "NR" аномальные ячейки. 

In [20]:
mask = df['rating'].str.contains('min', na=False)  # аналог WHERE rating LIKE '%min%' 
df.loc[mask, 'duration'] = df.loc[mask, 'rating']
df.loc[mask, 'rating'] = 'NR'

In [21]:
df['rating'] = df['rating'].fillna("NR")

### **3.5 date_added**

Колонку `date_added` мы разобьем на три: `day_added`, `month_added`, `year_added`.

"Unknown" вставим вместо пустых ячеек.

In [22]:
df['date_added'] = df['date_added'].str.strip() 
parsed = pd.to_datetime(df['date_added'], format='%B %d, %Y')  

df['day_added'] = parsed.dt.day.astype('Int64').astype(object)
df['month_added'] = parsed.dt.month.astype('Int64').astype(object)
df['year_added'] = parsed.dt.year.astype('Int64').astype(object)

mask_missing = df['date_added'].isna()
df.loc[mask_missing, ['day_added', 'month_added', 'year_added']] = 'Unknown'

df = df.drop(columns=['date_added'])

### **3.6 duration**

В данном случае колонку `duration` разобьем на две `duration_value` и `duration_unit`.

В пропуски заполним "Unknown".

In [23]:
splited_duration = df['duration'].str.split(' ', expand=True)
df['duration_value'] = splited_duration[0]
df['duration_unit'] = splited_duration[1]

df = df.drop(columns=['duration'])

## **Часть 4. Финальная проверка**

Здесь мы выполним те же команды, что и в части 2, для проверки.

In [24]:
print(f"Очищенный датафрейм содержит {df.shape[0]} строк и {df.shape[1]} колонок")

Очищенный датафрейм содержит 8807 строк и 15 колонок


In [25]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   show_id         8807 non-null   str   
 1   type            8807 non-null   str   
 2   title           8807 non-null   str   
 3   director        8807 non-null   str   
 4   cast            8807 non-null   str   
 5   country         8807 non-null   str   
 6   release_year    8807 non-null   int64 
 7   rating          8807 non-null   str   
 8   listed_in       8807 non-null   str   
 9   description     8807 non-null   str   
 10  day_added       8807 non-null   object
 11  month_added     8807 non-null   object
 12  year_added      8807 non-null   object
 13  duration_value  8807 non-null   str   
 14  duration_unit   8807 non-null   str   
dtypes: int64(1), object(3), str(11)
memory usage: 1.0+ MB


In [26]:
missing_after_cleaning = df.isna().sum()  
missing_after_cleaning.sort_values(ascending=False)

show_id           0
type              0
title             0
director          0
cast              0
country           0
release_year      0
rating            0
listed_in         0
description       0
day_added         0
month_added       0
year_added        0
duration_value    0
duration_unit     0
dtype: int64

In [27]:
round(missing_after_cleaning / len(df) * 100, 2).sort_values(ascending=False)

show_id           0.0
type              0.0
title             0.0
director          0.0
cast              0.0
country           0.0
release_year      0.0
rating            0.0
listed_in         0.0
description       0.0
day_added         0.0
month_added       0.0
year_added        0.0
duration_value    0.0
duration_unit     0.0
dtype: float64

In [28]:
if df.duplicated().sum() == 0:
    print("Полностью совпадающих строк нет")
else:
    print(f"Найдено {len(df[df.duplicated()])} дубликатов")
    df[df.duplicated()]

Полностью совпадающих строк нет


In [29]:
df.head(10)

,show_id,type,title,director,cast,country,release_year,rating,listed_in,description,day_added,month_added,year_added,duration_value,duration_unit
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,Unknown,United States,2020,PG-13,Documentaries,"As her father nears the end of his life, filmm...",25,9,2021,90,min
1,s2,TV Show,Blood & Water,Unknown,"Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...",South Africa,2021,TV-MA,"International TV Shows, TV Dramas, TV Mysteries","After crossing paths at a party, a Cape Town t...",24,9,2021,2,Seasons
2,s3,TV Show,Ganglands,Julien Leclercq,"Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...",Unknown,2021,TV-MA,"Crime TV Shows, International TV Shows, TV Act...",To protect his family from a powerful drug lor...,24,9,2021,1,Season
3,s4,TV Show,Jailbirds New Orleans,Unknown,Unknown,Unknown,2021,TV-MA,"Docuseries, Reality TV","Feuds, flirtations and toilet talk go down amo...",24,9,2021,1,Season
4,s5,TV Show,Kota Factory,Unknown,"Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...",India,2021,TV-MA,"International TV Shows, Romantic TV Shows, TV ...",In a city of coaching centers known to train I...,24,9,2021,2,Seasons
5,s6,TV Show,Midnight Mass,Mike Flanagan,"Kate Siegel, Zach Gilford, Hamish Linklater, H...",Unknown,2021,TV-MA,"TV Dramas, TV Horror, TV Mysteries",The arrival of a charismatic young priest brin...,24,9,2021,1,Season
6,s7,Movie,My Little Pony: A New Generation,"Robert Cullen, José Luis Ucha","Vanessa Hudgens, Kimiko Glenn, James Marsden, ...",Unknown,2021,PG,Children & Family Movies,Equestria's divided. But a bright-eyed hero be...,24,9,2021,91,min
7,s8,Movie,Sankofa,Haile Gerima,"Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...","United States, Ghana, Burkina Faso, United Kin...",1993,TV-MA,"Dramas, Independent Movies, International Movies","On a photo shoot in Ghana, an American model s...",24,9,2021,125,min
8,s9,TV Show,The Great British Baking Show,Andy Devonshire,"Mel Giedroyc, Sue Perkins, Mary Berry, Paul Ho...",United Kingdom,2021,TV-14,"British TV Shows, Reality TV",A talented batch of amateur bakers face off in...,24,9,2021,9,Seasons
9,s10,Movie,The Starling,Theodore Melfi,"Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...",United States,2021,PG-13,"Comedies, Dramas",A woman adjusting to life after a loss contend...,24,9,2021,104,min


In [30]:
df.describe()

,release_year
count,8807.000000
mean,2014.180198
std,8.819312
min,1925.000000
25%,2013.000000
50%,2017.000000
75%,2019.000000
max,2021.000000


In [31]:
columns = list(df.columns)
columns

['show_id',
 'type',
 'title',
 'director',
 'cast',
 'country',
 'release_year',
 'rating',
 'listed_in',
 'description',
 'day_added',
 'month_added',
 'year_added',
 'duration_value',
 'duration_unit']

## **Часть 6. Файл с очищенными данными**

In [32]:
df.to_csv('data/processed/netflix_titles_cleaned.csv', index=False)